File: Proyecto To Do List/tareas/modelo.go
package tareas

import (
	"time"
)

type EstadoTarea int

const (
	Activa EstadoTarea = iota
	Finalizada
	Suspendida
)

func (e EstadoTarea) String() string {
	switch e {
	case Activa:
		return "Activa"
	case Finalizada:
		return "Finalizada"
	case Suspendida:
		return "Suspendida"
	default:
		return "Desconocido"
	}
}

type Tarea struct {
	ID           int         `json:"id"`
	Titulo       string      `json:"titulo"`
	Estado       EstadoTarea `json:"estado"`
	Inicio       time.Time   `json:"inicio"`
	FinalizadaEn *time.Time  `json:"finalizada_en"`
}

type Tareas struct {
	Lista    []Tarea
	UltimoID int
}

func (t *Tareas) SetEstado(id int, nuevo EstadoTarea) error {
	index, err := t.buscarPorID(id)
	if err != nil {
		return err
	}

	tarea := &t.Lista[index]
	tarea.Estado = nuevo

	if nuevo == Finalizada {
		now := time.Now()
		tarea.FinalizadaEn = &now
	} else {
		tarea.FinalizadaEn = nil
	}

	return nil
}

File: Task Tracker Rust/src/tareas/modelo.rs
use chrono::{DateTime, Local};
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum EstadoTarea {
    Activa,
    Finalizada,
    Suspendida,
}

impl std::fmt::Display for EstadoTarea {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let estado = match self {
            EstadoTarea::Activa => "Activa",
            EstadoTarea::Finalizada => "Finalizada",
            EstadoTarea::Suspendida => "Suspendida",
        };
        write!(f, "{}", estado)
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Tarea {
    pub id: u32,
    pub titulo: String,
    pub estado: EstadoTarea,
    pub inicio: DateTime<Local>,
    pub finalizada_en: Option<DateTime<Local>>,
    pub subtareas: Vec<Tarea>,
    pub es_subtarea: bool,
    pub padre_id: Option<u32>,
    pub numero_jerarquico: String,
}

impl Tarea {
    pub fn todas_subtareas_completadas(&self) -> bool {
        if self.subtareas.is_empty() {
            return true;
        }
        self.subtareas.iter().all(|t| t.estado == EstadoTarea::Finalizada)
    }

    pub fn alguna_subtarea_activa(&self) -> bool {
        self.subtareas.iter().any(|t| t.estado == EstadoTarea::Activa)
    }

    pub fn alguna_subtarea_suspendida(&self) -> bool {
        self.subtareas.iter().any(|t| t.estado == EstadoTarea::Suspendida)
    }

    pub fn actualizar_estado_por_subtareas(&mut self) {
        if self.subtareas.is_empty() {
            return;
        }

        if self.todas_subtareas_completadas() {
            self.estado = EstadoTarea::Finalizada;
            self.finalizada_en = Some(Local::now());
        } else if self.alguna_subtarea_suspendida() {
            self.estado = EstadoTarea::Suspendida;
            self.finalizada_en = None;
        } else if self.alguna_subtarea_activa() {
            self.estado = EstadoTarea::Activa;
            self.finalizada_en = None;
        }
    }
}

#[derive(Debug, Default, Serialize, Deserialize)]
pub struct Tareas {
    pub lista: Vec<Tarea>,
    pub ultimo_id: u32,
    pub contador_principal: u32,
}